src/adapt/proto.ts

```diff
@@ -0,0 +1,499 @@
+    for (const key of Object.keys(nestedData.nested)) {
+      const data = nestedData.nested[key];
+      if (data) {
+        result.nested[key] = data;
+      }
+    }
+  }
+  return result;
+}
+
+function descriptorProtoFieldToField(
+  field: FieldDescriptorProto
+): protobuf.IField {
+  let type = '';
+  if (field.type) {
+    if (field.typeName) {
+      type = field.typeName;
+    } else {
+      type = fieldTypeLabelMap[field.type].replace('TYPE_', '').toLowerCase();
+    }
+  }
+  return {
+    id: field.number || -1,
+    type: type,
+    rule: labelToFieldRule(field.label),
+    options: optionToFieldOption(field.options),
+  };
+}
+
+function labelToFieldRule(label?: FieldDescriptorProto['label']): IFieldRule {
+  if (!label) {
+    return undefined;
+  }
+  return labelToFieldRuleMap[label];
+}
+
+function optionToFieldOption(
+  options: FieldDescriptorProto['options']
+): protobuf.IField['options'] {
+  if (!options) {
+    return {};
+  }
+  return {
+    ...options,
+  };
+}
+
+function convertTableFieldSchemaToFieldDescriptorProto(
+  field: TableFieldSchema,
+  fNumber: number,
+  scope: string,
+  useProto3: boolean
+): FieldDescriptorProto {
+  const name = `${field.name}`.toLowerCase();
+  const type = field.type;
+  if (!type) {
+    throw Error(`table field ${name} missing type`);
+  }
+  const label = convertModeToLabel(field.mode, useProto3);
+  let fdp: FieldDescriptorProto;
+  if (type === TableFieldSchema.Type.STRUCT) {
+    fdp = new FieldDescriptorProto({
+      name: name,
+      number: fNumber,
+      type: FieldDescriptorProto.Type.TYPE_MESSAGE,
+      typeName: scope,
+      label: label,
+    });
+  } else {
+    const pType = bqTypeToFieldTypeMap[type];
+    if (pType === null) {
+      throw Error(`table field type ${type} not supported`);
+    }
+    fdp = new FieldDescriptorProto({
+      name: field.name,
+      number: fNumber,
+      type: pType,
+      label: label,
+      options: {
+        packed: shouldPackType(pType, label, useProto3),
+      },
+      proto3Optional: isProto3Optional(label, useProto3),
+    });
+  }
+  return fdp;
+}
+
+function shouldPackType(
+  t: FieldDescriptorProtoType,
+  label: FieldDescriptorProtoLabel | null,
+  useProto3: boolean
+): boolean {
+  if (useProto3) {
+    return false;
+  }
+  if (label !== FieldDescriptorProto.Label.LABEL_REPEATED) {
+    return false;
+  }
+  return packedTypes.includes(t);
+}
+
+function isProto3Optional(
+  label: FieldDescriptorProtoLabel | null,
+  useProto3: boolean
+): boolean | undefined {
+  if (!useProto3) {
+    return undefined;
+  }
+  return label === FieldDescriptorProto.Label.LABEL_OPTIONAL;
+}
+
+function convertModeToLabel(
+  mode: TableFieldSchema['mode'],
+  useProto3: Boolean
+): FieldDescriptorProtoLabel | null {
+  if (!mode) {
+    return null;
+  }
+  return useProto3
+    ? bqModeToFieldLabelMapProto3[mode]
+    : bqModeToFieldLabelMapProto2[mode];
+}
```

Comment:
Overall comments, it might be good to go back and put in some jsdoc/tsdoc comments describing what the functions do. Or at least some basic non-doc comments. I'd also consider thinking about breaking this up into multiple files if you can, because it's sort of long.